### PR TITLE
Port max_unpool1d, max_unpool2d and max_unpool3d CUDA forward functions to ATen

### PR DESCRIPTION
**Pull request by andersenchen**
**Work in progress**

### CODE REVIEW

**Comment by ezyang** [comment]
> Hey Edward Yang! I implemented a version of maxunpool2d in CUDA here, and while it builds fine and runs tests fine on my machine and most others in the test suite, it fails on a number of instances.
>
> Question: 1) would you recommend a setup for debugging PyTorch, especially when it gets a bit tricky with GPU programming and other builds (e.g. IDE's, debuggers, profilers) and would I be able to gain access to the machines on which my PyTorch build has failed? I don't think I can replicate these errors otherwise.

First, I want to confirm that you actually ran the tests with CUDA enabled. You can determine this by running `python test/test_nn.py TestNN.test_MaxUnpool1d_net_cuda` and checking that the test reports actually as passed, as opposed to skipped.

I built your branch on my devgpu (on devgpu005.ash6) and I was able to reproduce the failure:

```
(/home/ezyang/local/pytorch-repro-env) [ezyang@devgpu005.ash6 ~/local/pytorch-repro] python test/test_nn.py TestNN.test_MaxUnpool1d_net_cuda
THCudaCheck FAIL file=/data/users/ezyang/pytorch-repro/aten/src/THC/THCTensorRandom.cu line=90 error=77 : an illegal memory access was encountered
E
======================================================================
ERROR: test_MaxUnpool1d_net_cuda (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/ezyang/pytorch-repro/test/common_utils.py", line 290, in wrapper
    method(*args, **kwargs)
  File "test/test_nn.py", line 6982, in <lambda>
    add(cuda_test_name, lambda self, test=test, kwargs=kwargs: test.test_cuda(self, **kwargs))
  File "/data/users/ezyang/pytorch-repro/test/common_nn.py", line 3067, in test_cuda
    self.test_noncontig(test_case, gpu_module, gpu_input)
  File "/data/users/ezyang/pytorch-repro/test/common_nn.py", line 2987, in test_noncontig
    out = test_case._forward(module, i)
  File "test/test_nn.py", line 499, in _forward
    return module(input)
  File "/home/ezyang/local/pytorch-repro-env/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/data/users/ezyang/pytorch-repro/test/common_utils.py", line 198, in freeze_rng_state
    torch.cuda.set_rng_state(cuda_rng_state)
  File "/data/users/ezyang/pytorch-repro/torch/cuda/random.py", line 49, in set_rng_state
    _lazy_call(cb)
  File "/data/users/ezyang/pytorch-repro/torch/cuda/__init__.py", line 122, in _lazy_call
    callable()
  File "/data/users/ezyang/pytorch-repro/torch/cuda/random.py", line 47, in cb
    _C._cuda_setRNGState(new_state_copy)
RuntimeError: cuda runtime error (77) : an illegal memory access was encountered at /data/users/ezyang/pytorch-repro/aten/src/THC/THCTensorRandom.cu:90

----------------------------------------------------------------------
Ran 1 test in 9.212s

FAILED (errors=1)
```

If you're following the fbcode instructions, chances are you are on a machine that does *not* have GPUs, and the test is being skipped. You need to run the test on a machine that actually has GPUs.

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+  auto* rawOutput = output.data<scalar_t>();
+
+  int maxp;
+  for (auto n = 0; n < numBatch; n++) {
+    auto nOffset = n * numChannels * outputWidth * outputHeight;
+    for (auto k = 0; k < numChannels; k++) {
+      auto finalOffset = nOffset + k * outputWidth * outputHeight;
+      auto* output_p_k = rawOutput + finalOffset;
+      auto* input_p_k = rawInput + finalOffset;
+      auto* ind_p_k = rawIndices + finalOffset;
```
Comment (ezyang) [inline aten/src/ATen/native/MaxUnpooling.cpp]: This, and the line above, look wrong. The original code looks like this:

```
    scalar_t *output_p_k = output_p + k*owidth*oheight;
    scalar_t *input_p_k = input_p + k*iwidth*iheight;
    THIndex_t *ind_p_k = ind_p + k*iwidth*iheight;
```

You're adding the input and the ind by output area, when you should add input area.

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+      for (auto i = 0; i < inputHeight; i++) {
+        for (auto j = 0; j < inputWidth; j++) {
+          maxp = ind_p_k[i * inputWidth + j];
```
Comment (ezyang) [inline aten/src/ATen/native/MaxUnpooling.cpp]: Getting rid of `TH_INDEX_BASE`: Good!

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+          if (maxp < 0 || maxp >= outputWidth * outputHeight) {
+            AT_ERROR("Invalid index");
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/native/MaxUnpooling.cpp]: I think if you put the `AT_ERROR` directly in here, you will thwart OpenMP (which is why you deleted it, I guess?) You should follow the pattern as the old code, using a `has_error` variable to test if an error has occurred, and don't throw an exception from inside the OpenMP parallelized code.

**Review by ezyang — CHANGES_REQUESTED** [review]
I did a quick review on one of your CPU kernels, and I noticed some transcription errors. I highly recommend that you put your new code and the old code side-by-side and double check that you didn't do any simple mistakes during the transcription :)

**Comment by andersenchen** [comment]
@ezyang thanks! I'm also on devgpu005.ash6 but was apparently running only the CPU tests.

File: aten/src/ATen/native/MaxUnpooling.cpp (hunk 2 above)
Comment (andersenchen) [reply aten/src/ATen/native/MaxUnpooling.cpp]: Thanks :)

**Comment by andersenchen** [comment]
Note that the latest CPU changes are here: https://github.com/pytorch/pytorch/pull/14845

In retrospect, I should've not stacked this PR on top of the CPU changes, and I should've removed those CPU commits from this PR; now that there are comments on the CPU code, though, I'll keep those commits here and rebase periodically.

**Comment by andersenchen** [comment]
See https://github.com/pytorch/pytorch/pull/15310 instead

**Comment by scopatz** [comment]
All open comments have been addressed in #19465

**Pull request closed without merging**